connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -181,7 +184,11 @@ public void execute() {
      */
     private boolean sendRecords() {
         int processed = 0;
-        for (final SourceRecord record : toSend) {
+        for (SourceRecord preTransformRecord : toSend) {
+            final SourceRecord record = transformationChain.apply(preTransformRecord);
+
+            if (record == null) continue;
```

Comment:
Your analysis makes sense. I think a HW can only be safely implemented by keeping track of uncommitted offsets and always picking the lowest one for the watermark (e.g. https://gist.github.com/shikhar/d4305ec18333c6623502). This is totally compatible with receiving `commitRecord()` for filtered records.